#Secant

---









#Question 6
(5 pts)
Find the minimizer of 
$$
f(x, y) = x^2 + xy + 10y^2 - 22y - 5x
$$
numerically by steepest descent.
1. For each iteration, record the values of $x,y$ and f and include in a table.
2. Plot the values on a contour plot.
3. Explore different starting values, such as (1, 10), (10, 10), (10, 1). Does the numbers of steps depend significantly on the starting guess?

In [ ]:
import numpy as np

In [ ]:
def secant(df, x0, x1, max_iter=100, tol=1e-3):
  for i in range(max_iter):
    y0 = df(x0)
    y1 = df(x1)

    x = (x0*y1 - x1*y0) / (y1 - y0) 
    y = df(x)

    if np.abs(x-x1) / np.abs(x1) < tol:
      break
    
    x0 = x1
    x1 = x
    y0 = y1
    y1 = y

  return x


In [ ]:
def steepest_descent(fun, grad, x0, max_iter=100, tol=1e-3):
  for i in range(max_iter):
    g = grad(x0)

    def d_ls_fun(a):
      return np.dot(g.T, grad(x0 - a*g))
      # test = np.dot(g.T, grad(x0 - a*g))
      # #print(f'test : {test}')
      # return test #line search function

    step_size = secant(d_ls_fun, 1, 2, 30, 1e-2)
    x = x0 - step_size * g #find the new point

    # print(f'Iteration:{i}, x1 = {x[0].item()}, x2 = {x[1].item()}, x3 = {x[2].item()}, function = {fun(x)}')
    print(f'Iteration:{i}, x1 = {x[0].item()}, x2 = {x[1].item()}, function = {fun(x)}')

    if np.linalg.norm(x-x0) / np.linalg.norm(x0) < tol:
      break

    x0 = x

  return x #optimal point

In [ ]:
def steepest_descent(fun, grad, x0, max_iter=100, tol=1e-3):
  for i in range(max_iter):
    g = grad(x0)

    # def d_ls_fun(a):
    #   return np.dot(g.T, grad(x0 - a*g))
      # test = np.dot(g.T, grad(x0 - a*g))
      # #print(f'test : {test}')
      # return test #line search function
    
    # step_size = secant(d_ls_fun, 1, 2, 30, 1e-2)

    # step_size = np.dot(g.T, grad(x0 - 0.5*g))
    step_size = 0.05
    x = x0 - step_size * g #find the new point

    # print(f'Iteration:{i}, x1 = {x[0].item()}, x2 = {x[1].item()}, x3 = {x[2].item()}, function = {fun(x)}')
    print(f'Iteration:{i}, x1 = {x[0].item()}, x2 = {x[1].item()}, function = {fun(x)}')

    if np.linalg.norm(x-x0) / np.linalg.norm(x0) < tol:
      break

    x0 = x

  return x #optimal point

In [ ]:
def fun(a):
  x = a[0].item()
  y = a[1].item()

  f_xy = x**2 + x*y + 10*y**2 - 22*y - 5*x 

  return f_xy

In [ ]:
def grad(a):
  x = a[0].item()
  y = a[1].item()


  g = np.zeros((2,1))
  g[0] = 2*x + y -5 #dfx
  g[0] = x + 20*y - 22 #dfy

  return g

#part

In [ ]:
def fun(x):
  x1 = x[0].item()
  x2 = x[1].item()
  x3 = x[2].item()
  y = (x1 - 4)**4 + (x2 - 3)**2 + 4*(x3 + 5)**4
  return y

In [ ]:
def grad(x):
  x1 = x[0].item()
  x2 = x[1].item()
  x3 = x[2].item()

  g = np.zeros((3, 1))
  g[0] = 4*((x1-4)**3)
  g[1] = 2 * (x2-3)
  g[2] = 16 * ((x3 + 5)**3)

  return g

In [ ]:
x0 = np.array([4, 2, -1])
x0 = np.reshape(x0, (3, 1))
x_test = steepest_descent(fun, grad, x0, 100)
print(x_test)

Iteration:0, x1 = 4.0, x2 = 2.0083625459975365, x3 = -5.281623550738581, function = 1.0085063020950835
Iteration:1, x1 = 4.0, x2 = 3.0015581821470687, x3 = -5.102654152637456, function = 0.00044661517419420136
Iteration:2, x1 = 4.0, x2 = 2.992862771702867, x3 = -5.054360348459882, function = 8.586925031224512e-05
Iteration:3, x1 = 4.0, x2 = 3.0002148935406554, x3 = -5.053036556345539, function = 3.169527175260388e-05
Iteration:4, x1 = 4.0, x2 = 2.9975337780027584, x3 = -5.038146096563194, function = 1.455180231208084e-05
Iteration:5, x1 = 4.0, x2 = 3.0000771103593, x3 = -5.03768815374993, function = 8.076055797505298e-06
[[ 4.        ]
 [ 3.00007711]
 [-5.03768815]]


In [ ]:
#choose start point at x0
x0 = np.array([1, 10])
x0 = np.reshape(x0, (2, 1))
x = steepest_descent(fun, grad, x0, 100)
print(x)

Iteration:0, x1 = -7.950000000000001, x2 = 10.0, function = 803.4525
Iteration:1, x1 = -16.4525, x2 = 10.0, function = 968.4222562500001
Iteration:2, x1 = -24.529875000000004, x2 = 10.0, function = 1259.0653925156253
Iteration:3, x1 = -32.20338125000001, x2 = 10.0, function = 1656.0408576828518
Iteration:4, x1 = -39.493212187500006, x2 = 10.0, function = 2142.2477479493987
Iteration:5, x1 = -46.418551578125005, x2 = 10.0, function = 2702.589172720426
Iteration:6, x1 = -52.99762399921875, x2 = 10.0, function = 3323.760029566474
Iteration:7, x1 = -59.247742799257814, x2 = 10.0, function = 3994.056312810717
Iteration:8, x1 = -65.18535565929493, x2 = 10.0, function = 4703.203814132299
Iteration:9, x1 = -70.82608787633018, x2 = 10.0, function = 5442.204284483993
Iteration:10, x1 = -76.18478348251367, x2 = 10.0, function = 6203.197316864919
Iteration:11, x1 = -81.27554430838798, x2 = 10.0, function = 6979.336381082797
Iteration:12, x1 = -86.11176709296858, x2 = 10.0, function = 7764.67759640

#Question 7
(9 pts)
Let the cost function of the unconstrained optimization problem of interest be:
$$
f(x) = \frac{1}{4}(x_{1} - 1)^2 + \sum^{n}_{i=2}(2x^{2}_{i-1} - x_{i} - 1)^2
$$
Consider two different scenarios, first $n=3$, and then $n=10$. Recall that the steepest descent algorithm is, 
$$
x^{k+1} = x^{k} - \alpha^{k} \nabla f(x^k), \alpha ^k \in \mathbb{R}_{>0}
$$
1. Using $x^0 = [-1.5,1,...,1]^T$ write out the first iteration of the steepest descent algorithm and obtain the optimum value for $\alpha^0$. \\
What is the value of $x^1$ if you implement $\alpha^0$? Verify that $f(x^1) < f(x^0).$
2. Write a code to find the minimizer of $f(x)$ using steepest descent algorithm with starting point of $x^0 = [-1.5, 1, ...,1]^T$ and using \\
$\bullet \alpha^k = argmin f(x^k - \alpha \nabla f(x^k))$ \\
$\bullet \text{constant } \alpha = 0.1$ \\
$\bullet \text{constant } \alpha = 0.5$ \\
$\bullet \text{constant } \alpha = 1.0$ (Use $||x^{k+1} - x^k|| \leq 10^{-6}$ as the stopping condition for your algorithm.)
3. How many steps does it take for the algorithm to converge for each choice of the step size above?
4. Use fiminbnd from Matlab or an equivalent function from the programming language of your choice to solve the problem. How many steps does it take for this algorithm to converge?
